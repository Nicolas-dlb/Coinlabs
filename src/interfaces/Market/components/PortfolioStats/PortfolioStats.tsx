import React, { useEffect, useState } from "react";
import "./PortfolioStats.scss";
import Graph from "components/Graph/Graph";
import Select from "components/Select/Select";
import { useSelector } from "react-redux";
import { selectHistory } from "redux/reducers/marketSlice";
import { pure } from "recompose";

type PortfolioStatsProps = {
  search?: string | undefined;
};
function PortfolioStats({ search }: PortfolioStatsProps) {
  const history = useSelector(selectHistory);

  const [active, setActive] = useState("Month");
  const [selectedCrypto, setSelectedCrypto] = useState("All");
  const [data, setData] = useState([{ tokens: 0, Price: 0 }]);

  useEffect(() => {
    if (search) {
      setSelectedCrypto(search);
    }
  }, [search]);

  const bitcoin = `bitcoin${active}`;
  const ethereum = `ethereum${active}`;
  const ripple = `ripple${active}`;
  const litecoin = `litecoin${active}`;
  const neo = `neo${active}`;

  const length =
    active === "Month"
      ? 601
      : active === "Week"
      ? 161
      : active === "Year"
      ? 361
      : active === "3Month"
      ? 81
      : active === "6Month"
      ? 181
      : 0;
  const multiplicateur = active === "Week" ? 0.94 : 1;
  const l = -length;
  useEffect(() => {
    if (
      history[bitcoin] &&
      history[ripple] &&
      history[litecoin] &&
      history[neo] &&
      history[ethereum]
    ) {
      setData(
        [...Array(length).keys()].map((x) => ({
          tokens: x,
          Price: history[bitcoin].slice(l)[x][1] * multiplicateur,
          ethereum: history[ethereum].slice(l)[x][1] * 11,
          ripple: history[ripple].slice(l)[x][1] * 5000,
          litecoin: history[litecoin].slice(l)[x][1] * 100,
          neo: history[neo].slice(l)[x][1] * 200,
        }))
      );
    } else {
      setData(
        [...Array(25).keys()].map((x) => ({
          tokens: x,
          Price: 0,
          ethereum: 0,
          ripple: 0,
          litecoin: 0,
          neo: 0,
          timestamp: 0,
        }))
      );
    }
  }, [history, active]);
  return (
    <div className="portfolio_stats">
      <p className="title">{search ? "Stats" : "Portfolio stats"}</p>
      <div className="portfolio_graph_container">
        <div className="tabs">
          <div className="tabs_left">
            <span
              tabIndex={0}
              role="button"
              onKeyDown={() => setActive("Week")}
              onClick={() => setActive("Week")}
              className={
                active === "Week"
                  ? "btn_portfolio btn_portfolio_active"
                  : "btn_portfolio"
              }
            >
              1w
            </span>
            <span
              role="button"
              tabIndex={0}
              onKeyDown={() => setActive("Month")}
              onClick={() => setActive("Month")}
              className={
                active === "Month"
                  ? "btn_portfolio btn_portfolio_active"
                  : "btn_portfolio"
              }
            >
              1m
            </span>
            <span
              role="button"
              tabIndex={0}
              onKeyDown={() => setActive("3Month")}
              onClick={() => setActive("3Month")}
              className={
                active === "3Month"
                  ? "btn_portfolio btn_portfolio_active"
                  : "btn_portfolio"
              }
            >
              3m
            </span>
            <span
              role="button"
              tabIndex={0}
              onKeyDown={() => setActive("6Month")}
              onClick={() => setActive("6Month")}
              className={
                active === "6Month"
                  ? "btn_portfolio btn_portfolio_active"
                  : "btn_portfolio"
              }
            >
              6m
            </span>
            <span
              role="button"
              tabIndex={0}
              onKeyDown={() => setActive("Year")}
              onClick={() => setActive("Year")}
              className={
                active === "Year"
                  ? "btn_portfolio btn_portfolio_active"
                  : "btn_portfolio"
              }
            >
              1y
            </span>
          </div>
          {!search && (
            <div
              role="button"
              tabIndex={0}
              onKeyDown={() => {
                document
                  .getElementById("select-c")!
                  .classList.toggle("toggle-select");
              }}
              id="select-c"
              onClick={() => {
                document
                  .getElementById("select-c")!
                  .classList.toggle("toggle-select");
              }}
              className="tabs_right"
            >
              <Select items="crypto" setCryptoSelected={setSelectedCrypto} />
            </div>
          )}
        </div>
        <Graph
          selected={selectedCrypto}
          data={data}
          color="Green"
          width={0}
          height={330}
          tooltip
        />
      </div>
    </div>
  );
}

export default pure(PortfolioStats);
