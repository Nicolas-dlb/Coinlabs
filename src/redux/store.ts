import { configureStore } from "@reduxjs/toolkit";
import marketReducer from "./reducers/marketSlice";
import walletsReducer from "./reducers/walletsSlice";
import appReducer from "./reducers/appSlice";
import userReducer from "./reducers/userSlice";

export default configureStore({
  reducer: {
    market: marketReducer,
    wallets: walletsReducer,
    app: appReducer,
    user: userReducer,
  },
});
