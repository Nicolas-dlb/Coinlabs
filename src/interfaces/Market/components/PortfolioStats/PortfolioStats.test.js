import React from "react";
import { mount } from "enzyme";
import store from "redux/store";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import PortfolioStats from "./PortfolioStats";

describe("Portfolio stats", () => {
  let PortfolioStatsComponent;
  beforeEach(() => {
    act(() => {
      PortfolioStatsComponent = mount(
        <Provider store={store}>
          <PortfolioStats />
        </Provider>
      );
    });
  });
  it("should render correctly", () => {
    expect(PortfolioStatsComponent).toMatchSnapshot();
  });
  it("should render Graph", () => {
    expect(PortfolioStatsComponent.find(".graph").first().exists()).toBe(true);
  });

  it("should render Select", () => {
    expect(PortfolioStatsComponent.find("#btn-select").first().exists()).toBe(
      true
    );
  });

  it("should render 5 button", () => {
    expect(PortfolioStatsComponent.find(".btn_portfolio").length).toBe(5);
  });
});
