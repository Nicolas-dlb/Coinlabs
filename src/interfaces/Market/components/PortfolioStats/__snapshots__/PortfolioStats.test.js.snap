// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio stats should render correctly 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <pure(PortfolioStats)>
    <PortfolioStats>
      <div
        className="portfolio_stats"
      >
        <p
          className="title"
        >
          Portfolio stats
        </p>
        <div
          className="portfolio_graph_container"
        >
          <div
            className="tabs"
          >
            <div
              className="tabs_left"
            >
              <span
                className="btn_portfolio"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                1w
              </span>
              <span
                className="btn_portfolio btn_portfolio_active"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                1m
              </span>
              <span
                className="btn_portfolio"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                3m
              </span>
              <span
                className="btn_portfolio"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                6m
              </span>
              <span
                className="btn_portfolio"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                1y
              </span>
            </div>
            <div
              className="tabs_right"
              id="select-c"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <Select
                items="crypto"
                setCryptoSelected={[Function]}
              >
                <div>
                  <select
                    className="crypto exchange_element"
                  >
                    <option
                      className="option exchange_element"
                      label="All"
                      value="All"
                    >
                      All
                    </option>
                    <option
                      className="option exchange_element"
                      label="Bitcoin"
                      value="Bitcoin"
                    >
                      Bitcoin
                    </option>
                    <option
                      className="option exchange_element"
                      label="Ethereum"
                      value="Ethereum"
                    >
                      Ethereum
                    </option>
                    <option
                      className="option exchange_element"
                      label="Ripple"
                      value="Ripple"
                    >
                      Ripple
                    </option>
                    <option
                      className="option exchange_element"
                      label="Litecoin"
                      value="Litecoin"
                    >
                      Litecoin
                    </option>
                    <option
                      className="option exchange_element"
                      label="Neo"
                      value="Neo"
                    >
                      Neo
                    </option>
                  </select>
                  <div
                    className="crypto-select exchange_element"
                  >
                    <svg
                      className="select-crypto-arrow exchange_element"
                      fill="#605b95"
                      height="24px"
                      viewBox="0 0 24 24"
                      width="24px"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0V0z"
                        fill="none"
                      />
                      <path
                        d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                      />
                    </svg>
                    <button
                      aria-label="btn-select"
                      className="btn-crypto exchange_element"
                      id="btn-select"
                      type="button"
                      value="All"
                    />
                    <div
                      className="crypto-list exchange_element"
                    >
                      <ul
                        id="crypto-items"
                      />
                    </div>
                  </div>
                </div>
              </Select>
            </div>
          </div>
          <Graph
            color="Green"
            data={
              Array [
                Object {
                  "Price": 0,
                  "tokens": 0,
                },
              ]
            }
            height={330}
            selected="All"
            tooltip={true}
            width={0}
          >
            <div
              className="graph_container"
            >
              <div
                className="graph"
              >
                <ResponsiveContainer
                  debounce={0}
                  height={330}
                  width="90%"
                >
                  <ResizeDetector
                    handleHeight={true}
                    handleWidth={true}
                    onResize={[Function]}
                    targetRef={
                      Object {
                        "current": <div
                          class="recharts-responsive-container"
                          style="width: 90%; height: 330px;"
                        />,
                      }
                    }
                  >
                    <div
                      className="recharts-responsive-container"
                      style={
                        Object {
                          "height": 330,
                          "maxHeight": undefined,
                          "minHeight": undefined,
                          "minWidth": undefined,
                          "width": "90%",
                        }
                      }
                    />
                  </ResizeDetector>
                </ResponsiveContainer>
                <Grid>
                  <div
                    className="grid"
                  >
                    <span
                      className="grid_card"
                    />
                    <span
                      className="grid_card"
                    />
                    <span
                      className="grid_card"
                    />
                    <span
                      className="grid_card"
                    />
                    <span
                      className="grid_card"
                    />
                  </div>
                </Grid>
              </div>
            </div>
          </Graph>
        </div>
      </div>
    </PortfolioStats>
  </pure(PortfolioStats)>
</Provider>
`;
